File: ios/runtime/XTRButton.m
//
//  XTRButton.m
//  XTSample
//
//  Created by 崔明辉 on 2017/8/29.
//  Copyright © 2017年 UED Center, YY Inc. All rights reserved.
//

#import "XTRButton.h"
#import "XTRUtils.h"
#import "XTRImage.h"
#import "XTRFont.h"
#import "XTRContext.h"
#import <XT-Mem/XTMemoryManager.h>

@interface XTRButton ()

@property (nonatomic, strong) UIButton *innerView;
@property (nonatomic, assign) BOOL vertical;
@property (nonatomic, assign) CGFloat inset;
@property (nonatomic, assign) BOOL highlighted;

@end

@implementation XTRButton

+ (NSString *)name {
    return @"XTRButton";
}

+ (NSString *)create {
    XTRButton *view = [[XTRButton alloc] initWithFrame:CGRectZero];
    view.innerView = [UIButton buttonWithType:UIButtonTypeSystem];
    view.innerView.adjustsImageWhenHighlighted = NO;
    [view.innerView setTitleColor:view.tintColor forState:UIControlStateNormal];
    [view.innerView addTarget:view action:@selector(onTouchUpInside) forControlEvents:UIControlEventTouchUpInside];
    [view.innerView addTarget:view action:@selector(onTouchStart) forControlEvents:UIControlEventTouchDown];
    [view.innerView addTarget:view action:@selector(onTouchEvent) forControlEvents:UIControlEventAllTouchEvents];
    [view addSubview:view.innerView];
    XTManagedObject *managedObject = [[XTManagedObject alloc] initWithObject:view];
    [XTMemoryManager add:managedObject];
    view.context = [JSContext currentContext];
    view.objectUUID = managedObject.objectUUID;
    return managedObject.objectUUID;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    self.innerView.frame = self.bounds;
}

+ (NSString *)xtr_title:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRButton class]]) {
        return [obj.innerView titleForState:UIControlStateNormal];
    }
    return nil;
}

+ (void)xtr_setTitle:(JSValue *)title objectRef:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRButton class]]) {
        [obj.innerView setTitle:[title toString] forState:UIControlStateNormal];
        [obj resetInset];
    }
}

+ (NSString *)xtr_font:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRButton class]]) {
        return [XTRFont create:[[obj.innerView titleLabel] font]];
    }
    return nil;
}

+ (void)xtr_setFont:(NSString *)fontRef objectRef:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    XTRFont *font = [XTMemoryManager find:fontRef];
    if ([obj isKindOfClass:[XTRButton class]] && [font isKindOfClass:[XTRFont class]]) {
        [obj.innerView.titleLabel setFont:font.innerObject];
    }
}

+ (NSString *)xtr_image:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRButton class]]) {
        XTRImage *image = (id)[obj.innerView imageForState:UIControlStateNormal];
        if ([image isKindOfClass:[XTRImage class]]) {
            return image.objectUUID;
        }
    }
    return nil;
}

+ (void)xtr_setImage:(NSString *)imageRef objectRef:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRButton class]]) {
        XTRImage *img = [XTMemoryManager find:imageRef];
        if ([img isKindOfClass:[XTRImage class]]) {
            [obj.innerView setImage:img.image forState:UIControlStateNormal];
            [obj resetInset];
        }
        else {
            [obj.innerView setImage:nil forState:UIControlStateNormal];
            [obj resetInset];
        }
    }
}

+ (NSDictionary *)xtr_color:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRButton class]]) {
        UIColor *color = [obj.innerView titleColorForState:UIControlStateNormal];
        if ([color isKindOfClass:[UIColor class]]) {
            return [JSValue fromColor:color];
        }
    }
    return @{};
}

+ (void)xtr_setColor:(JSValue *)color objectRef:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRButton class]]) {
        [obj.innerView setTitleColor:[color toColor] forState:UIControlStateNormal];
    }
}

+ (BOOL)xtr_vertical:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRButton class]]) {
        return obj.vertical;
    }
    return NO;
}

+ (void)xtr_setVertical:(BOOL)vertical objectRef:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRButton class]]) {
        obj.vertical = vertical;
    }
}

+ (CGFloat)xtr_inset:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRButton class]]) {
        return obj.inset;
    }
    return 0;
}

+ (void)xtr_setInset:(CGFloat)inset objectRef:(NSString *)objectRef {
    XTRButton *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRButton class]]) {
        obj.inset = inset;
    }
}

- (void)resetInset {
    if (self.vertical) {
        [self.innerView setImageEdgeInsets:UIEdgeInsetsMake(0, 0, 0, 0)];
        [self.innerView setTitleEdgeInsets:UIEdgeInsetsMake(0, 0, 0, 0)];
        [self layoutIfNeeded];
        self.innerView.imageEdgeInsets = UIEdgeInsetsMake(-(self.innerView.imageView.frame.size.height + self.innerView.titleLabel.frame.size.height + self.inset) / 2.0,
                                                          (self.innerView.frame.size.width / 2.0 - self.innerView.imageView.center.x) * 2,
                                                          0.0,
                                                          0.0);
        
        self.innerView.titleEdgeInsets = UIEdgeInsetsMake((self.innerView.imageView.frame.size.height + self.innerView.titleLabel.frame.size.height + self.inset) / 2.0,
                                                          (self.innerView.frame.size.width / 2.0 - self.innerView.titleLabel.center.x) * 2,
                                                          0.0,
                                                          0.0);
    }
    else {
        [self.innerView setImageEdgeInsets:UIEdgeInsetsMake(0, 0, 0, self.inset)];
        [self.innerView setTitleEdgeInsets:UIEdgeInsetsMake(0, self.inset, 0, 0)];
    }
}

- (void)onTouchUpInside {
    JSValue *value = self.scriptObject;
    if (value != nil) {
        [value invokeMethod:@"handleTouchUpInside" withArguments:@[]];
    }
}

- (void)onTouchStart {
    self.highlighted = NO;
}

- (void)onTouchEvent {
    if (self.highlighted != self.innerView.highlighted) {
        self.highlighted = self.innerView.highlighted;
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"handleHighlighted" withArguments:@[@(self.highlighted)]];
        }
        
    }
}

@end

File: android/xtruntime/src/main/java/com/opensource/xtruntime/XTRButton.kt
package com.opensource.xtruntime

import android.content.Context
import android.util.AttributeSet
import com.eclipsesource.v8.V8Object
import com.eclipsesource.v8.V8Value
import com.opensource.xtmem.XTManagedObject
import com.opensource.xtmem.XTMemoryManager

/**
 * Created by cuiminghui on 2017/9/8.
 */
class XTRButton @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : XTRView(context, attrs, defStyleAttr), XTRComponentInstance {

    var imageView: XTRImageView = XTMemoryManager.find(XTRImageView.create()) as XTRImageView
    var titleLabel: XTRLabel = XTMemoryManager.find(XTRLabel.create()) as XTRLabel

    init {
        XTRView.xtr_addSubview(imageView.objectUUID ?: "", this.objectUUID ?: "")
    }

    private var isVertical = false

    private var inset = 0.0

    override fun tintColorDidChange() {
        super.tintColorDidChange()
//        this.titleLabel.xtr_setTextColor(this.xtr_tintColorXTRTypes())
    }

    override fun layoutSubviews() {
        super.layoutSubviews()
        resetContents()
    }

    private fun resetContents() {
        val scale = resources.displayMetrics.density
        val imageWidth = (this.imageView.image?.size?.width ?: 0.0)
        val imageHeight = (this.imageView.image?.size?.height ?: 0.0)
        val textSize = this.titleLabel.intrinsicContentSize(this.width.toDouble() / scale  - imageWidth - inset) ?: XTRSize(0.0, 0.0)
        val textWidth = textSize.width
        val textHeight = textSize.height
        if (isVertical) {
            val contentHeight = imageHeight + inset + textHeight
            imageView.frame = XTRRect(
                    (this.bounds.width - imageWidth) / 2.0,
                    (this.bounds.height - contentHeight) / 2.0,
                    imageWidth,
                    imageHeight
            )
            titleLabel.frame = XTRRect(
                    (this.bounds.width - textWidth) / 2.0,
                    (this.bounds.height - contentHeight) / 2.0 + imageHeight + inset,
                    textWidth,
                    textHeight
            )
        }
        else {
            val contentWidth = imageWidth + inset + textWidth
            imageView.frame = XTRRect(
                    (this.bounds.width - contentWidth) / 2.0,
                    (this.bounds.height - imageHeight) / 2.0,
                    imageWidth,
                    imageHeight
            )
            titleLabel.frame = XTRRect(
                    (this.bounds.width - contentWidth) / 2.0 + imageWidth + inset,
                    (this.bounds.height - textHeight) / 2.0,
                    textWidth,
                    textHeight
            )
        }
    }

    companion object: XTRComponentExport() {

        override val name: String = "XTRButton"

        override fun exports(context: XTRContext): V8Object {
            val exports = V8Object(context.runtime)
            exports.registerJavaMethod(this, "create", "create", arrayOf())
            exports.registerJavaMethod(this, "xtr_title", "xtr_title", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_setTitle", "xtr_setTitle", arrayOf(String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_font", "xtr_font", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_setFont", "xtr_setFont", arrayOf(String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_image", "xtr_image", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_setImage", "xtr_setImage", arrayOf(String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_color", "xtr_color", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_setColor", "xtr_setColor", arrayOf(V8Object::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_vertical", "xtr_vertical", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_setVertical", "xtr_setVertical", arrayOf(Boolean::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_inset", "xtr_inset", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_setInset", "xtr_setInset", arrayOf(Double::class.java, String::class.java))
            return exports
        }

        fun create(): String {
            val view = XTRButton(XTRView.context.appContext)
            val managedObject = XTManagedObject(view)
            view.objectUUID = managedObject.objectUUID
            XTMemoryManager.add(managedObject)
            return managedObject.objectUUID
        }

        fun xtr_title(objectRef: String): String {
            return (XTMemoryManager.find(objectRef) as? XTRButton)?.titleLabel?.let {
                return@let XTRLabel.xtr_text(it.objectUUID ?: "")
            } ?: ""
        }

        fun xtr_setTitle(value: String, objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTRButton)?.titleLabel?.let {
                XTRLabel.xtr_setText(value, it.objectUUID ?: "")
            }
        }

        fun xtr_font(objectRef: String): String {
            return (XTMemoryManager.find(objectRef) as? XTRButton)?.titleLabel?.let {
                return@let XTRLabel.xtr_font(it.objectUUID ?: "")
            } ?: ""
        }

        fun xtr_setFont(fontRef: String, objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTRButton)?.titleLabel?.let {
                XTRLabel.xtr_setFont(fontRef, it.objectUUID ?: "")
            }
        }

        fun xtr_image(objectRef: String): String? {
            return (XTMemoryManager.find(objectRef) as? XTRButton)?.imageView?.let {
                return@let XTRImageView.xtr_image(it.objectUUID ?: "")
            } ?: null
        }

        fun xtr_setImage(imageRef: String, objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTRButton)?.let {
                XTRImageView.xtr_setImage(imageRef, it.imageView.objectUUID ?: "")
                it.resetContents()
            }
        }

        fun xtr_color(objectRef: String): V8Value {
            return XTRUtils.fromColor((XTMemoryManager.find(objectRef) as? XTRButton)?.tintColor ?: XTRColor(0.0, 0.0, 0.0, 0.0), context.runtime)
        }

        fun xtr_setColor(value: V8Object, objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTRButton)?.let {
                it.tintColor = XTRUtils.toColor(value)
            }
        }

        fun xtr_vertical(objectRef: String): Boolean {
            return (XTMemoryManager.find(objectRef) as? XTRButton)?.isVertical ?: false
        }

        fun xtr_setVertical(value: Boolean, objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTRButton)?.let {
                it.isVertical = value
                it.resetContents()
            }
        }

        fun xtr_inset(objectRef: String): Double {
            return (XTMemoryManager.find(objectRef) as? XTRButton)?.inset ?: 0.0
        }

        fun xtr_setInset(value: Double, objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTRButton)?.let {
                it.inset = value
                it.resetContents()
            }
        }

    }

}